<?xml version="1.0"?>
<launch>
    <!-- <args for urdf/> -->
    <arg name="camera_model"         default="OAK-D"       />
    <arg name="tf_prefix"            default="oak"         />
    <arg name="mode"                 default="depth"       />
    <arg name="base_frame"           default="oak-d_frame" />
    <arg name="parent_frame"         default="oak-d-base-frame" /> 
    <arg name="imuMode"              default="1" />   <!-- 0 -> COPY, 1 -> LINEAR_INTERPOLATE_GYRO, 2 -> LINEAR_INTERPOLATE_ACCEL -->

    <arg name="cam_pos_x"             default="0.0"        /> <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_pos_y"             default="0.0"        /> <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_pos_z"             default="0.0"        /> <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_roll"              default="0.0"        /> <!-- Orientation respect to base frame (i.e. "base_link) -->
    <arg name="cam_pitch"             default="0.0"        /> <!-- Orientation respect to base frame (i.e. "base_link) -->
    <arg name="cam_yaw"               default="0.0"        /> <!-- Orientation respect to base frame (i.e. "base_link) -->

    <arg name="lrcheck"               default="true"       /> 
    <arg name="extended"              default="false"      /> 
    <arg name="subpixel"              default="true"       />
    <arg name="rectify"               default="true"       />
    <arg name="depthAligned"          default="true"      />
     <!-- see https://docs.luxonis.com/projects/api/en/latest/components/nodes/stereo_depth/#currently-configurable-blocks for possible combination of optionf for higher fps-->
    <arg name="stereoFps"             default="15"         />
    <arg name="confidence"            default="200"        />
    <arg name="LRchecktresh"          default="5"          />
    <arg name="monoResolution"        default="720p"       /> 
    <arg name="angularVelCovariance"  default="0"       /> 
    <arg name="linearAccelCovariance" default="0"       /> 

    <include file="$(find depthai_bridge)/launch/urdf.launch">
        <arg name="base_frame"      value="$(arg  base_frame)"  />
        <arg name="parent_frame"    value="$(arg  parent_frame)"/>
        <arg name="camera_model"    value="$(arg  camera_model)"/>
        <arg name="tf_prefix"       value="$(arg  tf_prefix)"   />
        <arg name="cam_pos_x"       value="$(arg  cam_pos_x)"   />
        <arg name="cam_pos_y"       value="$(arg  cam_pos_y)"   />
        <arg name="cam_pos_z"       value="$(arg  cam_pos_z)"   />
        <arg name="cam_roll"        value="$(arg  cam_roll)"    />
        <arg name="cam_pitch"       value="$(arg  cam_pitch)"   />
        <arg name="cam_yaw"         value="$(arg  cam_yaw)"     />
    </include>

    <!-- launch-prefix="xterm -e gdb (add [- - args] without space) -->
    <node name="stereo_inertial_publisher" pkg="depthai_examples" type="stereo_inertial_node" output="screen" required="true">
        <param name="tf_prefix"             value="$(arg tf_prefix)"/>
        <param name="mode"                  value="$(arg mode)"/>
        <param name="imuMode"               value="$(arg imuMode)"/>
        <param name="lrcheck"               value="$(arg lrcheck)"/>
        <param name="extended"              value="$(arg extended)"/>
        <param name="subpixel"              value="$(arg subpixel)"/>
        <param name="rectify"               value="$(arg rectify)" />
        <param name="depthAligned"          value="$(arg depthAligned)" />
        <param name="stereoFps"             value="$(arg stereoFps)" />
        <param name="confidence"            value="$(arg confidence)" />
        <param name="LRchecktresh"          value="$(arg LRchecktresh)" />
        <param name="monoResolution"        value="$(arg monoResolution)" />
        <param name="angularVelCovariance"  value="$(arg angularVelCovariance)" />
        <param name="linearAccelCovariance" value="$(arg linearAccelCovariance)" />
        
        <param name="stereo/median_enable"                  value="true"/>
        <param name="stereo/speckle_enable"                 value="true"/>
        <param name="stereo/decimation_enable"              value="false"/>
        <param name="stereo/threshold_enable"               value="false"/>
        <param name="stereo/spatial_enable"                 value="false"/>

        <param name="stereo/median_mode"                    value="KERNEL_5x5"/>

        <param name="stereo/speckle_range"                  value="50"/>

        <param name="stereo/temporal_enable"                value="false"/>
        <param name="stereo/temporal_mode"                  value="VALID_2_IN_LAST_4"/>
        <param name="stereo/temporal_alpha"                 value="0.4"/>
        <param name="stereo/temporal_delta"                 value="20"/>

        <param name="stereo/spatial_radius"                 value="2"/>
        <param name="stereo/spatial_alpha"                  value="0.5"/>
        <param name="stereo/spatial_delta"                  value="0"/>
        <param name="stereo/spatial_iterations"             value="1"/>

        <param name="stereo/threshold_min"                  value="200"/>
        <param name="stereo/threshold_max"                  value="30000"/>
        <param name="stereo/decimation_mode"                value="1"/>
        <param name="stereo/decimation_factor"              value="1"/>
        
    </node>            

    <node pkg="nodelet" type="nodelet" name="nodelet_manager"  args="manager" output="screen"/>

    <node pkg="nodelet" type="nodelet" name="depth_image_convertion_nodelet"
        args="load depth_image_proc/convert_metric nodelet_manager">
        <remap from="image_raw" to="/stereo_inertial_publisher/stereo/depth"/>    
        <remap from="image" to="/stereo_inertial_publisher/stereo/image"/>
    </node>


    <node if="$(eval arg('depthAligned') == true)" pkg="nodelet" type="nodelet" name="depth_image_to_rgb_pointcloud"
        args="load depth_image_proc/point_cloud_xyzrgb nodelet_manager">
        <remap from="rgb/camera_info" to="/stereo_inertial_publisher/color/camera_info"/>
        <remap from="rgb/image_rect_color" to="/stereo_inertial_publisher/color/image"/>
        <remap from="depth_registered/image_rect" to="/stereo_inertial_publisher/stereo/image"/>    
        <remap from="depth_registered/points" to="/stereo_inertial_publisher/stereo/points"/>
    </node>

    <node if="$(eval arg('depthAligned') == false)" pkg="nodelet" type="nodelet" name="depth_image_to_pointcloud"
        args="load depth_image_proc/point_cloud_xyz nodelet_manager">
        <remap from="camera_info" to="/stereo_inertial_publisher/stereo/camera_info"/>
        <remap from="image_rect" to="/stereo_inertial_publisher/stereo/image"/>    
        <remap from="points" to="/stereo_inertial_publisher/stereo/points"/>
    </node>


  <node if="$(eval arg('depthAligned') == true)" type="rviz" name="rviz" pkg="rviz" args="-d $(find depthai_examples)/rviz/stereoInertialDepthAlign.rviz" />
  <node if="$(eval arg('depthAligned') == false)" type="rviz" name="rviz" pkg="rviz" args="-d $(find depthai_examples)/rviz/stereoInertialROS1.rviz" />

</launch>

